boxes/helper: Add support for rendering tables.

preetmishra opened the pull request:
I have added a helper function, render_table, to process a table and return its content as a list of strings which can then be rendered in the MessageBox and amended the test.

This implementation is v1. Feedbacks are welcomed. :) 
There can certainly be follow-ups to improve this further.

Partially fixes #452.

@neiljp

kaustubh-nair commented:
@preetmishra This is a great feature to have :tada: 
I've tested this locally a little bit, and it seems to work well.

It would be useful to be able to support the syntax used by the web-app
If I use the example given [here](https://zulipchat.com/help/format-your-message-using-markdown#tables), I run into an `IndexError` exception.

preetmishra commented:
@kaustubh-nair Thanks for the review. :+1:  I have sorted out the bug that you found.

preetmishra commented:
Another thing I have noticed that in-table styling doesn't work in ZT yet (for instance, alignment). I was thinking that we might want to support that in v2 but if it doesn't seem like a good idea, I would move ahead to support that in v1 itself.

preetmishra commented:
@neiljp Thanks for the review. I have made the changes that you suggested for v1.

I have:
* Removed all the horizontal lines (except the top, middle and bottom border).
* Used Unicode characters for drawing the table which makes it cleaner overall.
* Added support for alignments.
* Refactored the code quite a bit (including splitting the previous render function into two - `parse_html_table` and `render_table`).


I have left comments in the code for explaining a few points, let me know if it is too much or if we need more. :)

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1397,7 +1397,15 @@ def test_private_message_to_self(self, mocker):
+         '<tbody><tr><td>John</td><td>Doe</td></tr><tr><td>Mary</td><td>Moe'
+         '</td></tr></tbody></table>', [
+            '┌───────────┬──────────┐\n',
+            ('bold', '│ Firstname │ Lastname │\n'),
```
I'm unsure about the best render, but it would seem better to me to have just the cell contents in `bold` or some other head-style, rather than the box parts too? This is particularly evident in the blue and light themes, where I first tested this.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1397,7 +1397,15 @@ def test_private_message_to_self(self, mocker):
+            '├───────────┼──────────┤\n',
+            '│ John      │ Doe      │\n',
+            '│ Mary      │ Moe      │\n',
+            '└───────────┴──────────┘']),
```
You only have one test here, but support at least different alignments, of which there is no indication here?

neiljp commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,95 @@ def notify(title: str, html_text: str) -> None:
+    """
+    headers = table_element.thead.tr.find_all('th')
+    rows = table_element.tbody.find_all('tr')
+    alignments = []
```
This is on a column-basis, right? So maybe `column_alignments`?

preetmishra replied:
Yes, `column_alignments` is a better choice.

preetmishra replied:
Yes, it on a column-basis, `column_alignments` will be perfect.

preetmishra replied:
Yes, it is on a column-basis, column_alignments will be perfect.

preetmishra replied:
Yes, it is on a column-basis, `column_alignments` will be perfect.

neiljp commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,95 @@ def notify(title: str, html_text: str) -> None:
+    ]
+
+    top_border = unicode_border(u'┌', u'─', u'┬', u'┐', column_widths)
+    middle_border = unicode_border(u'├', u'─', u'┼', u'┤', column_widths)
```
I like the way you've extracted the box-drawing elements here :+1: 

You might like to do the same for the vertical bars?

preetmishra replied:
I have added a `vertical_bar` to store the Unicode character to make that cleaner. I believe the way we have it right now is okay, doing it this way will make things complicated.

preetmishra replied:
Thanks. :+1: 

I have added a `vertical_bar` to store the Unicode character to make that cleaner. I believe the way we have it right now is okay, doing it this way will make things complicated.

neiljp commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,95 @@ def notify(title: str, html_text: str) -> None:
+            table.append(top_border)
+            # Make the header bold.
+            table.append(('bold', styled_row))
+            table.append(middle_border)
```
Would this be clearer as two sections, one generating the header and the other the rest (the latter being the loop)?

neiljp commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,95 @@ def notify(title: str, html_text: str) -> None:
+    return (alignments, cells)
+
+
+def unicode_border(lcorner: str, line: str, connector: str, rcorner: str,
```
Is this more accurately a `unicode_row`? Or `row_with_only_borders`?

This would make it clearer when you interleave it with a `row_with_styled_content`, perhaps.

preetmishra replied:
I agree, these will be more accurate.

neiljp commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,95 @@ def notify(title: str, html_text: str) -> None:
+    border = lcorner
+    for width in widths:
+        # Add +2 to the width to compensate for the extra space added around
+        # each string while styling the row by `style_row`.
```
It would be simpler to centralize this padding convention, rather than have it over two methods?

neiljp commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,95 @@ def notify(title: str, html_text: str) -> None:
+    # [[row0, row0, row0],
+    # [row1, row1, row1],
+    # [row2, row2, row2]].
+    # Add +1 to count the header row as well.
```
This looks like it's part of the comment above? I assume it is not?

Perhaps the cell description should go into the docstring?

preetmishra replied:
No, they are different. I'll move the upper part into the docstring.

preetmishra commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,141 @@ def notify(title: str, html_text: str) -> None:
+    fill = fill_char * fill_width
+
+    # Pad the leftmost box-drawing character.
+    row_strip[0] = row_strip[0] + fill  # type: ignore
```
`mypy` complains about this concatenation.

preetmishra commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,141 @@ def notify(title: str, html_text: str) -> None:
+    fill = fill_char * fill_width
+
+    # Pad the leftmost box-drawing character.
+    row_strip[0] = row_strip[0] + fill  # type: ignore
```
I am not sure why `mypy` complains about these concatenations.

preetmishra replied:
I tried using `type: str` for `fill` but that didn't seem to solve the issue. I'll add comments regarding the ignores.

preetmishra commented:
@neiljp Thanks. I have added extra test cases (to support different alignments), fixed the styling issue and incorporated the other changes that you suggested.

I have restructured the row styling functions to return a list of strings and tuples (style, string). This not only helps in centralizing the padding routine but also helps in setting styles individually. In addition to these, this also makes the work required for adding inline-table-styles in v2 rather straightforward. 

Once we reach a consensus about the implementation, I would add the tests for the helper functions.

preetmishra commented:
@neiljp Thanks. I have added extra test cases (to support different alignments), fixed the styling issue and incorporated the other changes that you suggested.

I have restructured the row styling functions to return a list of strings and tuples (style, string). This not only helps in centralizing the padding routine but also helps in setting styles individually. In addition to these, this also makes the work required for adding inline-table-styles in v2 rather straightforward. 

I would add the tests for the helper functions once we reach a consensus about the implementation.

preetmishra commented:
Update: Extended the docstring for `pad_row_strip`.

neiljp commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,142 @@ def notify(title: str, html_text: str) -> None:
+
+
+def pad_row_strip(row_strip: TABLE_ROW, fill_char: str=' ',
+                  fill_width: int=1) -> TABLE_ROW:
```
I like this new method :+1:

neiljp commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,142 @@ def notify(title: str, html_text: str) -> None:
+    for row in cells:
+        table.extend(row_with_styled_content(row, column_alignments,
+                                             column_widths))
+    table.extend(bottom_border)
```
This layout for constructing the table content looks cleaner to me :+1:

sumanthvrao commented:
@neiljp This is the TinglingGit Output for All the PR's which clash with the file ui_tools/boxes.py
and their diff
```
PR#397(+2-0)
PR#530(+4-0)
PR#278(+16-8)
PR#539(+1-1)
PR#540(+63-22)
PR#542(+26-18)
PR#287(+51-1)
PR#424(+8-1)
PR#572(+38-38)
PR#573(+11-10)
PR#335(+18-10)
PR#354(+2-0)
PR#550(+4-3)
PR#359(+12-1)
PR#371(+21-15)
PR#501(+6-2)
```

preetmishra commented on zulipterminal/helper.py:
```diff
@@ -493,3 +493,142 @@ def notify(title: str, html_text: str) -> None:
+                                         newline=False)
+
+    # Construct the table, row-by-row.
+    table = []  # type: List[Union[str, Tuple[Optional[str], str]]]
```
Yes, this seems reasonable. I think `StyleTableData` will be good.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1397,7 +1397,41 @@ def test_private_message_to_self(self, mocker):
+            '│ ', (None, 'Robert'), ' │ ', (None,  '  1'), ' │\n',
+            '│ ', (None, ' Mary '), ' │ ', (None, '100'), ' │\n',
+            '└─', '──────', '─┴─', '───', '─┘',
+         ]),
```
The tables will have at least two rows (the header and an empty body row) if nothing is defined in the body.

Though it is possible for a table to have one column. I'll add a few more test cases.

preetmishra commented:
@neiljp Thanks. I have added a comment for the first commit (it lets `row_with_styled_content` add `None` display attribute) and added a few more test cases for the table.

I have replaced `TABLE_ROW` with `StyledTableData` as it seemed more reasonable. On a related note, `urwid_Size` type alias uses underscore with CamelCase, was this intentional or should we refactor?

The `vertical_bar` is now a function argument for `row_with_styled_data`. This would make replacing box-drawing characters easy, if we ever feel like to.

If it is not much of an overhead, we can move the MessageBox in a separate file along with the new changes and perhaps separate tests for the table from `soup2mark` tests.

preetmishra commented:
@neiljp Thanks. I have:

* Added a comment for the first commit (it lets `row_with_styled_content` add `None` display attribute).
* Added a few more test cases for the table.
* Replaced `TABLE_ROW` with `StyledTableData` as it seemed more reasonable. On a related note, `urwid_Size` type alias uses underscore with CamelCase, was this intentional or should we refactor?
* The `vertical_bar` is now a function argument for `row_with_styled_data`. This would make replacing box-drawing characters easy if we ever feel like to.
* Added comments for the `type: ignore` issues. I tried using a `type: str` for `fill` but it didn't seem to work.

If it is not much of an overhead, we can move the MessageBox in a separate file along with the new changes and perhaps separate the table test from `soup2mark` tests (thoughts?).

preetmishra commented:
@neiljp Thanks. I understand the idea for `urwid_Size` now.

I was able to fix all the `mypy` issues using `cast`, which assists `mypy` with the type, and what you suggested in https://github.com/zulip/zulip-terminal/pull/550#discussion_r403626995 :+1:.

neiljp commented:
@preetmishra As I mentioned on the stream, this is now merged with a few minor tweaks :tada: This is really great :) I'm sure you can find it in the list of commits by now!

preetmishra commented:
@neiljp Thanks so much, it was fun working on this. :)

To address the things that we would like to see in the future iterations, I'll file an issue.
